p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,336 @@
+	}
+	c.m.Lock()
+	defer c.m.Unlock()
+	for k, stream := range c.streams {
+		// reset the streams, but we do not need to be notified
+		// of stream closure
+		stream.close(true, false)
+		delete(c.streams, k)
+	}
+
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{
+		Transport: "webrtc",
+	}
+}
+
+// Close closes the underlying peerconnection.
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	c.cancel()
+	return c.pc.Close()
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	streamID, err := c.idAllocator.nextID()
+	if err != nil {
+		return nil, err
+	}
+	dc, err := c.pc.CreateDataChannel("", &webrtc.DataChannelInit{ID: streamID})
+	if err != nil {
+		return nil, err
+	}
+	rwc, err := c.detachChannel(ctx, dc)
+	if err != nil {
+		return nil, fmt.Errorf("open stream: %w", err)
+	}
+	stream := newStream(
+		c,
+		dc,
+		rwc,
+		nil,
+		nil,
+	)
+	c.addStream(stream)
+	return stream, nil
+}
+
+func (c *connection) AcceptStream() (network.MuxedStream, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, os.ErrClosed
+	case accStream := <-c.acceptQueue:
+		stream := newStream(
+			c,
+			accStream.channel,
+			accStream.stream,
+			nil,
+			nil,
+		)
+		c.addStream(stream)
+		return stream, nil
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+func (c *connection) LocalPrivateKey() ic.PrivKey {
+	return c.privKey
+}
+
+func (c *connection) RemotePeer() peer.ID {
+	return c.remotePeer
+}
+
+func (c *connection) RemotePublicKey() ic.PubKey {
+	return c.remoteKey
+}
+
+// implement network.ConnMultiaddrs
+func (c *connection) LocalMultiaddr() ma.Multiaddr {
+	return c.localMultiaddr
+}
+
+func (c *connection) RemoteMultiaddr() ma.Multiaddr {
+	return c.remoteMultiaddr
+}
+
+// implement network.ConnScoper
+func (c *connection) Scope() network.ConnScope {
+	return c.scope
+}
+
+func (c *connection) Transport() tpt.Transport {
+	return c.transport
+}
+
+func (c *connection) addStream(stream *webRTCStream) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.streams[stream.id] = stream
+}
+
+func (c *connection) removeStream(id uint16) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	delete(c.streams, id)
+}
+
+func (c *connection) onConnectionStateChange(state webrtc.PeerConnectionState) {
+	log.Debugf("[%s][%d] handling peerconnection state: %s", c.localPeer, c.dbgId, state)
+	if state == webrtc.PeerConnectionStateFailed || state == webrtc.PeerConnectionStateClosed {
+		// reset any streams
+		c.resetStreams()
+		c.cancel()
+		c.scope.Done()
+		c.pc.Close()
+	}
+}
+
+// detachChannel detaches an outgoing channel by taking into account the context
```

Comment:
will see to explain it better

Reply:
The underlying SCTP stream for a datachannel implements a `net.Conn` interface. However, the datachannel creates a goroutine which continuously reads from the SCTP stream and surfaces the data using an `OnMessage` callback. The actual abstractions are as follows: `webrtc.DataChannel` wraps `pion.DataChannel`, which wraps `sctp.Stream`.

The goroutine for reading, `Detach` method, and the `OnMessage` callback are present at the `webrtc.DataChannel` level. `Detach` provides us abstracted access to the underlying `pion.DataChannel`, which allows us to issue `Read` calls to the datachannel. This was desired because it was not feasible to introduce backpressure with the `OnMessage` callbacks. The tradeoff is a change in the semantics of the `OnOpen` callback, and having to force close `Read` locally.